### PR TITLE
CORE - SLO Processing Rules Verifications

### PR DESCRIPTION
**Pull request by blen-desta**
#### What does this PR do (if it's not clear from the Title)? Please include any specification references that apply.
- 3.2.1 - Complex Type RequestAbstractType
- 3.7.3.2 - Session Authority Rules

Changes related to this PR are made on https://github.com/codice/ddf/pull/3283. 

#### Checklist:
- [x] SAML Spec Table of Contents documentation updated
- [x] Unit Tests Added/Modified

### CODE REVIEW

File: ctk/common/src/main/kotlin/org/codice/compliance/verification/core/responses/CoreLogoutResponseProtocolVerifier.kt [hunk 1]
```diff
@@ -13,11 +13,33 @@
...
+import org.codice.compliance.utils.TestCommon.Companion.SUCCESS
 import org.codice.compliance.verification.core.ResponseVerifier
 import org.codice.security.saml.SamlProtocol
 import org.opensaml.saml.saml2.core.LogoutRequest
 import org.w3c.dom.Node
 
 class CoreLogoutResponseProtocolVerifier(logoutRequest: LogoutRequest, samlResponseDom: Node,
-    binding: SamlProtocol.Binding) :
-    ResponseVerifier(logoutRequest, samlResponseDom, binding)
+    binding: SamlProtocol.Binding, private val expectedStatusCode: List<String> = listOf(SUCCESS)) :
```
Comment (brjeter) [inline ctk/common/src/main/kotlin/org/codice/compliance/verification/core/responses/CoreLogoutResponseProtocolVerifier.kt]: ✏️ `expectedStatusCodes` with an "s"

File: ctk/common/src/main/kotlin/org/codice/compliance/verification/core/responses/CoreLogoutResponseProtocolVerifier.kt [hunk 2]
```diff
@@ -13,11 +13,33 @@
...
+    ResponseVerifier(logoutRequest, samlResponseDom, binding) {
+
+    override fun verify() {
+        super.verify()
+        verifyStatusCode()
+    }
+
+    private fun verifyStatusCode() {
+        if (samlResponseDom.recursiveChildren(STATUS_CODE)
+                .none { expectedStatusCode.contains(it.attributeText("Value")) })
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/verification/core/responses/CoreLogoutResponseProtocolVerifier.kt]: The top-level status code is covered in `ResponseVerifier`
```
if (samlResponseDom.recursiveChildren("Status")
          .any {
                    !topLevelStatusCodes
                              .contains(it.children(STATUS_CODE).first()
                                        .attributeText("Value"))
          })
```

File: ctk/common/src/main/kotlin/org/codice/compliance/verification/core/requests/CoreLogoutRequestProtocolVerifier.kt [hunk 3]
```diff
@@ -13,14 +13,17 @@
...
 import org.codice.compliance.attributeNode
 import org.codice.compliance.verification.core.CommonDataTypeVerifier
 import org.codice.compliance.verification.core.RequestVerifier
+import org.codice.security.saml.SamlProtocol
 import org.w3c.dom.Node
 
-class CoreLogoutRequestProtocolVerifier(private val samlRequestDom: Node) :
-    RequestVerifier(samlRequestDom) {
+class CoreLogoutRequestProtocolVerifier(private val samlRequestDom: Node,
+    binding: SamlProtocol.Binding) : RequestVerifier(samlRequestDom, binding) {
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/verification/core/requests/CoreLogoutRequestProtocolVerifier.kt]: We're not using binding in this class so I didn't add `val` before it. It needs to be a `val` to be `private`.

File: ctk/common/src/main/kotlin/org/codice/compliance/verification/core/responses/CoreLogoutResponseProtocolVerifier.kt [hunk 4]
```diff
@@ -13,11 +13,37 @@
...
+    override fun verify() {
+        super.verify()
+        verifySecondaryStatusCode()
+    }
+
+    private fun verifySecondaryStatusCode() {
+        if (expectedStatusCode != null && samlResponseDom.children("Status")
+                .flatMap { it.children(STATUS_CODE) }
+                .firstOrNull()
+                ?.children(STATUS_CODE)
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/verification/core/responses/CoreLogoutResponseProtocolVerifier.kt]: Yes, ☝️ is the structure

**Comment by brjeter** [comment]
Successfully hero'd ✅

**Pull request merged**